### PR TITLE
MDC auto-configuration for Brave and OpenTelemetry

### PR DESCRIPTION
**Pull request by marcingrzejszczak**
A sample with a test that proves that things are working fine [mdc-observability.zip](https://github.com/spring-projects/spring-boot/files/9471155/mdc-observability.zip)

### CODE REVIEW

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/OpenTelemetryConfigurations.java [hunk 1]
```diff
@@ -168,6 +187,139 @@ OtelHttpServerHandler otelHttpServerHandler(OpenTelemetry openTelemetry) {
...
+
+				@Bean
+				@Conditional(BaggageTagSpanHandlerCondition.class)
+				BaggageTaggingSpanProcessor baggageTaggingSpanProcessor(TracingProperties properties) {
+					return new BaggageTaggingSpanProcessor(properties.getBaggage().getTagFields());
+				}
+
+				/**
+				 * We need a special condition as it users could use either comma or yaml encoding,
+				 * possibly with a deprecated prefix.
```
Comment (wilkinsona) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/OpenTelemetryConfigurations.java:244]: Why could there be a deprecated prefix?

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/BraveAutoConfiguration.java [hunk 2]
```diff
@@ -166,6 +186,137 @@ BraveHttpClientHandler braveHttpClientHandler(
...
+						return SpanHandler.NOOP; // Brave ignores these
+					}
+					return new BaggageTagSpanHandler(tagFields.stream().map(BaggageField::create).toArray(BaggageField[]::new));
+				}
+
+			}
 
+			/**
+			 * We need a special condition as it users could use either comma or yaml encoding,
+			 * possibly with a deprecated prefix.
```
Comment (wilkinsona) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/BraveAutoConfiguration.java:255]: Why could there be a deprecated prefix?

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/BraveAutoConfiguration.java [hunk 3]
```diff
@@ -166,6 +186,137 @@ BraveHttpClientHandler braveHttpClientHandler(
 			return new BraveHttpClientHandler(httpClientHandler);
 		}
 
-	}
+		@Configuration(proxyBeanMethods = false)
+		// TODO: Add this to JSON
```
Comment (wilkinsona) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/BraveAutoConfiguration.java:190]: Why can't this be added to the additional metadata now?

File: build.gradle [hunk 4]
```diff
@@ -25,6 +25,7 @@ allprojects {
 	group "org.springframework.boot"
 
 	repositories {
+		mavenLocal() // for Tracing snapshots
```
Comment (wilkinsona) [inline build.gradle:28]: We can't merge this as it'll fail on CI. The plan is to switch to snapshots for Tracing and Micrometer tomorrow so this will shortly become unnecessary anyway.

File: spring-boot-project/spring-boot-actuator-autoconfigure/build.gradle [hunk 5]
```diff
@@ -51,6 +51,7 @@ dependencies {
 	optional("io.micrometer:micrometer-core")
 	optional("io.micrometer:micrometer-tracing")
 	optional("io.micrometer:micrometer-tracing-bridge-brave")
+	optional("io.zipkin.aws:brave-propagation-aws:0.23.4") // TODO: How should we align to this version?
```
Comment (wilkinsona) [inline spring-boot-project/spring-boot-actuator-autoconfigure/build.gradle:54]: We've got something support for keeping versions aligned in `spring-boot-dependencies`. Please add dependency management for all of the modules published from https://github.com/openzipkin/zipkin-aws to `spring-boot-dependencies`, remove the version declaration here, and we can take it from there.

**Review by wilkinsona — CHANGES_REQUESTED** [review]

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/OpenTelemetryConfigurations.java (hunk 1 above)
Comment (marcingrzejszczak) [reply spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/OpenTelemetryConfigurations.java:239]: Most likely that's some ancient thing that I unnecessarily copied over from Sleuth. I'll remove it

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/BraveAutoConfiguration.java (hunk 2 above)
Comment (marcingrzejszczak) [reply spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/BraveAutoConfiguration.java:262]: Most likely that's some ancient thing that I unnecessarily copied over from Sleuth. I'll remove it

File: build.gradle (hunk 4 above)
Comment (marcingrzejszczak) [reply build.gradle]: Removed

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/OpenTelemetryAutoConfiguration.java [hunk 6]
```diff
@@ -34,6 +35,7 @@
  */
 @AutoConfiguration(before = MicrometerTracingAutoConfiguration.class)
 @Import({ SdkConfiguration.class, TracerConfiguration.class, MicrometerConfiguration.class })
+@EnableConfigurationProperties(TracingProperties.class)
```
Comment (wilkinsona) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/OpenTelemetryAutoConfiguration.java:38]: Why is this needed? It's duplicating the same from `TracerConfiguration`.

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/BraveAutoConfiguration.java [hunk 7]
```diff
@@ -141,8 +169,8 @@ static class BraveMicrometer {
 
 		@Bean
 		@ConditionalOnMissingBean
-		BraveTracer braveTracerBridge(brave.Tracer tracer, CurrentTraceContext currentTraceContext,
-				BraveBaggageManager braveBaggageManager) {
+		@Primary // TODO: Tracer implements BaggageManager but there are 2 BaggageManager beans (Tracer and BraveBaggageManager)
```
Comment (wilkinsona) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/BraveAutoConfiguration.java:172]: Is there a plan to address this? We won't merge something with a TODO.

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/OpenTelemetryConfigurations.java [hunk 8]
```diff
@@ -105,10 +120,9 @@ Sampler otelSampler(TracingProperties properties) {
 		}
 
 		@Bean
-		@ConditionalOnMissingBean
 		SpanProcessor otelSpanProcessor(List<SpanExporter> spanExporter) {
 			return SpanProcessor.composite(spanExporter.stream()
-					.map((exporter) -> BatchSpanProcessor.builder(exporter).build()).collect(Collectors.toList()));
+					.map(exporter -> BatchSpanProcessor.builder(exporter).build()).collect(Collectors.toList()));
```
Comment (wilkinsona) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/OpenTelemetryConfigurations.java:125]: `(exporter) ->` is Spring Boot's code style and this doesn't pass Checkstyle. Please build your changes before submitting them.

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/OpenTelemetryConfigurations.java [hunk 9]
```diff
@@ -168,6 +182,105 @@ OtelHttpServerHandler otelHttpServerHandler(OpenTelemetry openTelemetry) {
 					new DefaultHttpServerAttributesExtractor());
 		}
 
+		@Configuration(proxyBeanMethods = false)
+		static class PropagationConfiguration {
+
+			@Bean
+			@ConditionalOnMissingBean
+			@ConditionalOnProperty(value = "management.tracing.propagation", havingValue = "B3", matchIfMissing = true)
+			TextMapPropagator b3TextMapPropagator() {
```
Comment (wilkinsona) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/OpenTelemetryConfigurations.java:191]: To provide as much type information as possible to the bean factory, the signature should show that it returns a `B3Propagator`.

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/OpenTelemetryConfigurations.java [hunk 10]
```diff
@@ -168,6 +182,105 @@ OtelHttpServerHandler otelHttpServerHandler(OpenTelemetry openTelemetry) {
...
+				return B3Propagator.injectingSingleHeader();
+			}
+
+			@Configuration(proxyBeanMethods = false)
+			@ConditionalOnProperty(value = "management.tracing.baggage.enabled", havingValue = "false")
+			static class NoBaggagePropagatorConfiguration {
+				@Bean
+				@ConditionalOnMissingBean
+				@ConditionalOnProperty(value = "management.tracing.propagation", havingValue = "W3C")
+				TextMapPropagator w3cTextMapPropagator() {
```
Comment (wilkinsona) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/OpenTelemetryConfigurations.java:201]: To provide as much type information as possible to the bean factory, the signature should show that it returns a `W3CTraceContextPropagator`.

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/OpenTelemetryConfigurations.java [hunk 11]
```diff
@@ -168,6 +182,105 @@ OtelHttpServerHandler otelHttpServerHandler(OpenTelemetry openTelemetry) {
...
+					return W3CTraceContextPropagator.getInstance();
+				}
+
+			}
+
+			@Bean
+			@ConditionalOnProperty(value = "management.tracing.propagation", havingValue = "AWS")
+			@ConditionalOnClass(AWSPropagation.class)
+			TextMapPropagator awsTextMapPropagator() {
```
Comment (wilkinsona) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/OpenTelemetryConfigurations.java:210]: To provide as much type information as possible to the bean factory, the signature should show that it returns an `AwsXrayPropagator`.

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/OpenTelemetrySlf4jApplicationListener.java [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+
+package org.springframework.boot.actuate.autoconfigure.tracing;
+
+import io.micrometer.tracing.otel.bridge.Slf4JEventListener;
+
+import org.springframework.context.ApplicationEvent;
+import org.springframework.context.ApplicationListener;
+import org.springframework.context.PayloadApplicationEvent;
+
+// TODO: Should this live here?
```
Comment (wilkinsona) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/OpenTelemetrySlf4jApplicationListener.java:25]: If it's going to live in Boot, it probably belongs in `spring-boot-actuator` and this module just auto-configures it.

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/OpenTelemetrySlf4jBaggageApplicationListener.java [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+
+import java.util.List;
+
+import io.micrometer.tracing.otel.bridge.Slf4JBaggageEventListener;
+
+import org.springframework.context.ApplicationEvent;
+import org.springframework.context.ApplicationListener;
+import org.springframework.context.PayloadApplicationEvent;
+
+// TODO: Should this live here?
```
Comment (wilkinsona) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/OpenTelemetrySlf4jBaggageApplicationListener.java:27]: If it's going to live in Boot, it probably belongs in spring-boot-actuator and this module just auto-configures it.

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/TracingEnvironmentPostProcessor.java [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+import org.springframework.core.env.MutablePropertySources;
+import org.springframework.core.env.PropertySource;
+
+/**
+ * Adds default properties for the application: logging pattern level that prints trace
+ * information (e.g. trace ids).
+ *
+ * @author Dave Syer
+ * @author Marcin Grzejszczak
+ * @since 3.0.0
```
Comment (wilkinsona) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/TracingEnvironmentPostProcessor.java:35]: `@since` isn't needed as the class is package-private.

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/TracingEnvironmentPostProcessor.java [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+
+	static final String PROPERTY_SOURCE_NAME = "defaultProperties";
+
+	@Override
+	public void postProcessEnvironment(ConfigurableEnvironment environment, SpringApplication application) {
+		Map<String, Object> map = new HashMap<String, Object>();
+		// This doesn't work with all logging systems but it's a useful default so you see
+		// traces in logs without having to configure it.
+		if (Boolean.parseBoolean(environment.getProperty("management.tracing.enabled", "true"))) {
+			// TODO: Add management.tracing.default-logging-pattern-enabled as a property in the json
```
Comment (wilkinsona) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/TracingEnvironmentPostProcessor.java:47]: Please add this to the json and remove the TODO.

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/TracingEnvironmentPostProcessor.java [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+ * Adds default properties for the application: logging pattern level that prints trace
+ * information (e.g. trace ids).
+ *
+ * @author Dave Syer
+ * @author Marcin Grzejszczak
+ * @since 3.0.0
+ */
+class TracingEnvironmentPostProcessor implements EnvironmentPostProcessor {
+
+	static final String PROPERTY_SOURCE_NAME = "defaultProperties";
```
Comment (wilkinsona) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/TracingEnvironmentPostProcessor.java:39]: I'm not sure that this should mess with the `defaultProperties` property source. It may be better for it to add its own property source so that it's more clear in things like the `env` endpoint where the properties have come from.

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/TracingProperties.java [hunk 17]
```diff
@@ -53,4 +74,119 @@ public void setProbability(float probability) {
 
 	}
 
+	public static class Baggage {
+
+		/**
+		 * Enables correlating the baggage context with logging contexts.
```
Comment (wilkinsona) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/TracingProperties.java:80]: For consistency with other enabled properties, this should be something like "Whether to enable correlation of the baggage context with logging contexts"

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/TracingProperties.java [hunk 18]
```diff
@@ -53,4 +74,119 @@ public void setProbability(float probability) {
...
+
+	}
+
+	public static class Propagation {
+
+		/**
+		 * Tracing context propagation types.
+		 */
+		// TODO: Should this be the default?
+		private PropagationType type = PropagationType.B3;
```
Comment (wilkinsona) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/TracingProperties.java:154]: This default should be declared in the additional metadata as the annotation processor cannot detect default enum values.

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/TracingProperties.java [hunk 19]
```diff
@@ -53,4 +74,119 @@ public void setProbability(float probability) {
...
+
+			/**
+			 * W3C propagation type.
+			 */
+			W3C,
+
+			/**
+			 * Jaeger propagation type.
+			 */
+			JAEGER, // TODO: Supported by OTel, not by Brave - my suggestion is not to support it
```
Comment (wilkinsona) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/TracingProperties.java:183]: We could either not support it or fail on startup if someone sets the value to `jaeger` but they're using Brave. I don't know the space well enough to have an opinion about which would be best. If you're suggesting that we don't support it, that's good enough for me and this could be removed.

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/TracingProperties.java [hunk 20]
```diff
@@ -53,4 +74,119 @@ public void setProbability(float probability) {
...
+
+			/**
+			 * Custom propagation type. If picked, requires bean registration overriding the
+			 * default propagation mechanisms.
+			 */
+			CUSTOM
```
Comment (wilkinsona) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/TracingProperties.java:189]: Should we switch off the default registrations if custom is used? I wonder if that would be easier to diagnose than the defaults being used accidentally.

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/zipkin/ZipkinProperties.java [hunk 21]
```diff
@@ -44,6 +44,11 @@ public class ZipkinProperties {
 	 */
 	private Duration readTimeout = Duration.ofSeconds(10);
 
+	/**
+	 * Name of this application registered in Zipkin. Defaults to {@code spring.application.name}.
```
Comment (wilkinsona) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/zipkin/ZipkinProperties.java:48]: We can't use javadoc tags in property descriptions.

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/test/java/org/springframework/boot/actuate/autoconfigure/tracing/BaggageAutoConfigurationTests.java [hunk 22]
```diff
@@ -0,0 +1,195 @@
...
+
+import org.springframework.boot.autoconfigure.AutoConfigurations;
+import org.springframework.boot.test.context.runner.ApplicationContextRunner;
+import org.springframework.context.ApplicationContext;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.mock;
+
+/**
+ * Tests for {@link BraveAutoConfiguration}.
```
Comment (wilkinsona) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/test/java/org/springframework/boot/actuate/autoconfigure/tracing/BaggageAutoConfigurationTests.java:40]: Should be `BaggageAutoConfiguration`

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/test/java/org/springframework/boot/actuate/autoconfigure/tracing/BaggageAutoConfigurationTests.java [hunk 23]
```diff
@@ -0,0 +1,195 @@
...
+ *
+ * @author Moritz Halbritter
```
Comment (wilkinsona) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/test/java/org/springframework/boot/actuate/autoconfigure/tracing/BaggageAutoConfigurationTests.java:42]: Wrong author?

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/test/java/org/springframework/boot/actuate/autoconfigure/tracing/OpenTelemetryConfigurationsMicrometerConfigurationTests.java [hunk 24]
```diff
@@ -99,6 +101,12 @@ void shouldBackOffOnCustomBeans() {
 		});
 	}
 
+	@Configuration(proxyBeanMethods = false)
+	@EnableConfigurationProperties(TracingProperties.class)
```
Comment (wilkinsona) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/test/java/org/springframework/boot/actuate/autoconfigure/tracing/OpenTelemetryConfigurationsMicrometerConfigurationTests.java:105]: This is a bit unusual. Why is a user configuration class enabling one of our configuration properties classes?

**Review by wilkinsona — CHANGES_REQUESTED** [review]

**Comment by wilkinsona** [comment]
Sorry, looks like GitHub has double-posted my latest review comments.

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/OpenTelemetryAutoConfiguration.java (hunk 6 above)
Comment (marcingrzejszczak) [reply spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/OpenTelemetryAutoConfiguration.java:38]: `TracerConfiguration` doesn't have this. After this change it looks exactly the same way as `BraveAutoConfiguration` looks like

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/test/java/org/springframework/boot/actuate/autoconfigure/tracing/OpenTelemetryConfigurationsMicrometerConfigurationTests.java (hunk 24 above)
Comment (marcingrzejszczak) [reply spring-boot-project/spring-boot-actuator-autoconfigure/src/test/java/org/springframework/boot/actuate/autoconfigure/tracing/OpenTelemetryConfigurationsMicrometerConfigurationTests.java:105]: That's because Moritz has created a test for a subconfiguration only that now requires the TracingProperties and those are enabled in a higher order configuration.

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/OpenTelemetrySlf4jBaggageApplicationListener.java (hunk 13 above)
Comment (marcingrzejszczak) [reply spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/OpenTelemetrySlf4jBaggageApplicationListener.java]: We worked it around to not include these classes at all

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/OpenTelemetrySlf4jApplicationListener.java (hunk 12 above)
Comment (marcingrzejszczak) [reply spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/OpenTelemetrySlf4jApplicationListener.java]: We worked it around to not include these classes

**Comment by marcingrzejszczak** [comment]
closing in favour of https://github.com/spring-projects/spring-boot/pull/32480
